### PR TITLE
[move-lang][move-ir] Intern all the strings

### PR DESCRIPTION
**Pull request by tnowacki**
- Use move-symbol-pool's Symbol for all strings in the IR and source language ASTs
- After fixing everything, I hit a wall of clippy errors. about 75% of the way through, I realized I should have added the `Copy` trait to these items in a separate PR. So sorry about that... luckily the changes are mostly mechanical. 

## Motivation

Should reduce the memory overhead of the Move compiler (greatly if Rust can't optimize this well). But I have not gotten memory profiling working yet. 

Runtime perf in debug builds was a wash. But increased perf by about 30% in release builds!
Unfortunately, the overhead of interning is an issue with `nextest` as they each test operates in their own binary (I've been told).
```
dev build DF+Std (averaged over 10 iterations)
OLD       3.058s
INTERNED  2.833s
REDUCTION -7.36%

release build DF+Std (averaged over 100 iterations)
OLD       0.292s
INTERNED  0.205s
REDUCTION -29.8%

nextest -p move-lang
OLD       29.937s
INTERNED  33.230s
INCREASE  +11.00%

nextest -p move-lang-functional-tests
OLD       210.312s
INTERNED  237.315s
INCREASE  +12.840%

nextest -p move-prover
OLD       458.118s
INTERNED  480.420s
INCREASE  +4.5812%

x test -p move-lang
OLD       8.615s
INTERNED  8.541s
-         -

x test -p move-lang-functional-tests
OLD       44.905s
INTERNED  44.897s
-         -

x test -p move-prover
OLD       332.042s
INTERNED  317.463s
REDUCTION -4.3907%
```

## Test Plan

- Ran tests
- Looked at perf

### CODE REVIEW

File: language/move-symbol-pool/src/pool.rs [hunk 1]
```diff
@@ -69,7 +69,7 @@ impl Pool {
     /// which top-level bucket contains an entry corresponding to the string,
     /// as well as a scalar value that can be used to quickly check whether
     /// two strings are not equal.
-    fn hash(string: &str) -> u64 {
+    fn hash(string: &Cow<str>) -> u64 {
```
Comment (modocache) [inline language/move-symbol-pool/src/pool.rs:72]: Just to make sure I understand, this is to avoid a clone that occurs when creating a hash for the `insert` method? I guess it could be split out into its own pull request, but whatever (GitHub makes it so annoying to split things up and stack them...).

File: language/move-model/src/model.rs [hunk 2]
```diff
@@ -54,7 +54,7 @@ use move_binary_format::{
 use move_core_types::{
     account_address::AccountAddress, identifier::Identifier, language_storage, value::MoveValue,
 };
-use move_symbol_pool::Symbol as StringSymbol;
+use move_symbol_pool::Symbol as MoveStringSymbol;
```
Comment (modocache) [inline language/move-model/src/model.rs:57]: Cool, this is a better name I think 👍

**Review by modocache — DISMISSED** [review]
Phew! A truly massive change 🐘

I'm very curious how many strings end up in the pool, and how much memory it ends up consuming, when compiling some of the larger Move programs in the test suite. Is it anything to worry about?

File: language/move-symbol-pool/src/pool.rs (hunk 1 above)
Comment (tnowacki) [reply language/move-symbol-pool/src/pool.rs:72]: There shouldn't be a change here. This was in another PR that I thought had already landed 

> GitHub makes it so annoying to split things up and stack them...
lol it is why my PRs end up being so big a lot of times :P

File: language/move-model/src/model.rs (hunk 2 above)
Comment (tnowacki) [reply language/move-model/src/model.rs:57]: Yeah I thought it was more clear as it started being used more

**Comment by tnowacki** [comment]
> I'm very curious how many strings end up in the pool, and how much memory it ends up consuming, when compiling some of the larger Move programs in the test suite. Is it anything to worry about?

It is a good question. I have not gotten memory profiling to work on macOS but my expectations would be as follows:
* For a typical program in our codebase today, there are no deps/libs passed in, so all Move code is being compiled. So I would expect in this case the memory overhead to be strictly less. Even for a large program. Since I would presume that most strings are used more than once (most strings/identifiers probably have a declaration+ a usage), interning would cut the the memory overhead from strings by about half
* In the worst case, if a bunch of deps/libs were passed in with really big functions, we would intern all of the identifiers in those Move functions, but then drop the functions for compiler optimization reasons (the dep function bodies don't need to be compiled). So all of those strings didn't need to be interned. But those strings will be allocated always currently for parsing. So the peak memory usage will remain the same for the deps/libs, but it will sustain for longer. I don't think this is a big issue because...
* The most common use case in the future will be with the package system. In that case the deps/libs will be populated from CompiledModules which go through the "interface generator". That interface generator does not generate function bodies. So the previous issue is moot.

TLDR
* All source programs (no deps) will have strictly less memory usage
* deps/libs might be an issue of over interning but...
* the interface generator makes the common future usage of deps/libs not an issue

**Comment by tnowacki** [comment]
Based on running `time -l` I got a rough estimate that this saves about 27% of peak memory usage!
```
/usr/bin/time -l -- cargo run --bin move-build -- -a=Std=0x1 -a=DiemFramework=0x1 -a=DiemRoot=0xA550C18 -a=CurrencyInfo=0xA550C18 -a=TreasuryCompliance=0xB1E55ED -a=VMReserved=0x0 -- language/diem-framework/modules language/move-stdlib/modules
``` 
Before
```
33189888  peak memory footprint
```
After
```
24088576  peak memory footprint
```

**Review by modocache — DISMISSED** [review]
Wow! Incredible, thanks for checking!

**Review by modocache — APPROVED** [review]
